### ISSUE TITLE
PHP Interned strings

### ISSUE THREAD

**Issue by SirSnyder** [krakjoe/pthreads#802]
Interned string are thread local and therefore problematic. Pthreads shares interned strings between thread contexts. As a simple example zend_class_entry->name. 

Thread A creates a zend_class_entry(ce) with interned zend_string as name. Thread B copies this ce via ```pthreads_copy_entry```. ce->name is apparently copied(```zend_string_new```), but it isn't, cause the string is interned. Thread A shuts down, internal strings are freed. Access to ce->name in Thread B ends with a seg fault.

Macro ```zend_string_new``` mapped to function ```zend_string_dup(zend_string *s, int persistent)```. In case of an interned string, ```zend_string_dup``` returns the passed zend_string pointer and no duplicated string, as expected.

Any idea how to solve this issue?

**Comment by tpunt** [comment]
I thought pthreads delayed the joining of a parent thread until the child thread has also been joined? At least that's what I am seeing with a script such as:
```php
<?php

class T extends Thread
{
    public function run()
    {
        $t = new class extends Thread {
            public function run()
            {
                while (1); // blocks php_request_shutdown for thread 2
            }
        };

        $t->start();
        // thread shutdown blocked by nested thread 3
    }
}

$t = new T;

$t->start();
```

There could be something wrong with the delaying logic on thread shutdown (either it is not being performed or needs to be prolonged). Otherwise, if this really is a problem, then the simple solution would be to just perform a hard copy of all interned strings with something such as:
```c
zend_string_init(ZSTR_VAL(interned_str), ZSTR_LEN(interned_str), GC_FLAGS(interned_str) & IS_STR_PERSISTENT);
```

Do you have a script to reproduce the segfault with the interned class names example you gave?

**Comment by SirSnyder** [comment]
The relation between parent and child thread is okay. The other way, it's about interned strings defined in child thread and copied into parent thread.

Example, a separate file(`ExternalBaseClass.php`) with 

```
class ExternalBaseClass extends \Threaded {}
```

and the script

```
class Foo extends \Thread {
	/** @var bool */
	public $running;

	/** @var \Threaded */
	private $shared;

	public function __construct(\Threaded $shared) {
		$this->shared = $shared;
	}

	public function run() {
		$this->running = true;

		require __DIR__ .'/ExternalBaseClass.php';

		$this->shared['baseClass'] = new ExternalBaseClass();

		$this->synchronized(function () {
			while($this->running) {
				$this->wait(0);
			}
		});
	}
}
$shared = new \Threaded();

$foo = new Foo($shared);
$foo->start();

while(true) {
	if(!isset($shared['baseClass'])) {
		continue;
	}
	// copy zend_class_entry
	$baseClass = $shared['baseClass'];

	break;
}
$foo->running = false;
$foo->notify() && $foo->join();

// die; // seg fault in zend_string_release()

$baseClass = $shared['baseClass'];

var_dump('BaseClass');
var_dump($baseClass); // seg fault in zend_string_copy()
```

My fix to hard copy interned strings was ugly in contrast to your one-liner. I've pushed the possible fix in a separate branch. A following issue is a memory leak but I have no idea where this leak comes from repectively why the parent thread probably doesn't free the hard copied string.

**Comment by tpunt** [comment]
Whilst this is certainly a problem, it does seem like a fairly self-contained edge case. I don't think we perform a copy of a child class entry (or any other structure that may contain interned strings) from a child thread to a parent thread anywhere else.

I therefore don't think we need to copy *all* interned strings. That just feels like a waste of time and memory in most cases. A possible fix would be to set another ZG (such as `hard_copy_interned_strings`), where it is set to `0` by default, and then toggled it to `1` for the `pthreads_prepared_entry` invocation that occurs in the object connection code only. Interning the strings should fix the memory leak, too.

I've pushed a potential fix in the [interned-strings2](https://github.com/krakjoe/pthreads/compare/interned-strings2) branch - let me know if you can see any issues with it.

**Comment by SirSnyder** [comment]
I've run some tests and I'm fine with the fix, no issues seen. :-)

**Comment by tpunt** [comment]
@SirSnyder Thanks for checking.

I decided to give it another review, since I did not like how everything was just being interned. After some digging, it turned out the memory leak was occurring on the class filename being copied. I had already fixed filename copying for op arrays using a Zend global, so I reused that fix here again. Things seem to work fine with [the new patch](https://github.com/krakjoe/pthreads/compare/interned-strings2), and it feels cleaner too. I've also added a slightly simplified version of your test too.

The only thing that springs to mind that could be problematic is that we have removed some string copying logic (such as calls to `zend_string_new`) where we knew string interning would occur. I think that is going to come back to bite us with this child-to-parent copying edge case. Quite likely, we will have to go through the copying logic to make sure any string copying makes use of `zend_string_new`, to take into account the hard copying of interned strings.

**Comment by tpunt** [comment]
Ok, I've gone through the copying logic and have identified the following 3 cases which may be problematic:
 - `prepare_class_property_table`: `info->name` (see 536a7b29ddd57a25209b5fb79b5abcf0ecd85f79)
 - `pthreads_copy_literals`: `IS_STRING` types (see 5aebd199ee71b9fdfc61618574a45db9e63e97f5)
 - `pthreads_copy_internal_function`: `function->common.function_name` (see f7d147ca4042d59c038a8455f59ce8ecadb3ecdb)

I tried to trigger the interned strings problem with these cases, but I couldn't. So for now, I'm going to simply merge the new patch as-is. Should we see problems in future around these cases, then the above commits can be individually reverted to resolve such issues.

**Comment by tpunt** [comment]
Merged in 88c915adef07e78e1e423c39dadb05ce165330f6.
